### ISSUE TITLE
3.x is now in beta

### ISSUE THREAD

**Issue by stancl** [archtechx/tenancy#397]
Version 3.x of Tenancy for Laravel is now in **open beta**.

3.x is mostly a ground-up rewrite. It's focused on making the package **much more flexible** by using an **event-based architecture** and it brings **"enterprise" features** such as synced resources (e.g. users) between tenant databases, user impersonation, easy Cashier integration, and others.

The documentation is being rewritten completely, and soon also the website will be rebuilt completely.
 
## Added

(All of the previous features, including mutli-database tenancy and automatic tenancy mode are still part of the package.)

- [x] **Shared resources** between tenant databases (e.g. users)
- [x] **Single-database tenancy** traits
- [x] **Manual** mode of tenancy (model traits etc)
- [x] **Rich event system**
- [x] Job pipelines
- [x] Path identification
- [x] Subdomain identification
- [x] Native **Eloquent support** -- tenants are now Eloquent models (easier Cashier, Nova, etc integration)
- [x] **User impersonation**
- [x] Nova resources for managing tenants
- [x] **Fully featured example SaaS application to see how the package is used in production**

Join our Discord to share feedback, get more updates and join the beta: https://discord.gg/7cpgPxv

**Comment by mikeerickson** [comment]
@stancl two questions 

1. Does your approach use a different database for each tenant 
2. And if not, how can one get into the beta program 

I am at this phase of our application design and am looking at all avenues.  But don’t want to go the unique DB for each tenant route

**Comment by stancl** [comment]
Hi.

The package supports both multi-database tenancy and single-database tenancy.

You can join the beta on our Discord: https://discord.gg/7cpgPxv

**Comment by mikeerickson** [comment]
Thank you, I will he getting started with this tomorrow!

**Comment by vahidalvandi** [comment]
Hi , When does it become stable?

**Comment by stancl** [comment]
2 weeks or so

**Comment by waltergalvao** [comment]
Hi @stancl 

I am starting a new project, and this package seems nice. Should I wait for the v3 release? I like the idea of single database tenancy

Uh.. will the documentation will very different from v2? Are there many breaking changes?

Thanks!

**Comment by stancl** [comment]
Hi.

The documentation will be very different. And so will be the usage of the package.

However note that since the package bootstraps tenancy automatically, the only bits of code that you have to change are the ones where you interact with the package logic directly - creating tenants etc.

A stableish beta should be ready sometime this week.

**Comment by lostlian** [comment]
> Hi.
> 
> The documentation will be very different. And so will be the usage of the package.
> 
> However note that since the package bootstraps tenancy automatically, the only bits of code that you have to change are the ones where you interact with the package logic directly - creating tenants etc.
> 
> A stableish beta should be ready sometime this week.

So there will be an upgrade guide comming in the documentation?

**Comment by stancl** [comment]
Yes

**Comment by creati0ns** [comment]
Hi,

Am currently in the process of testing various tenancy packages. What i really do miss here, is tenant specific views out of the box. Surely it can be implemented without out-of-the-box support for it, but still, since this packages is also compared with others in docs..

Consider this use-case.
Am building a multi site, which should serve 3 base websites, out of the same code base, with different features. Only one of these 3 base websites, should be offering SaaS based (sub) domains.
In this scenario it would've helped a lot to already have tenant specific views implemented.

**Comment by stancl** [comment]
So you want tenant-specific Blade files? That's not true "multi-tenancy" though, you're manually changing the code because of specific customers.

I think using non-executable (= not blade, but twig or liquid) views, letting your customers change them as they want, and storing the processed version in a Redis cache or something like that is a far superior option. But this is very opinionated, tenant-specific views mean a different thing to every project, and it's hard for me to add a one-size-fits-all solution. I do want out-of-the-box functionality, but not for things that will just make the package bigger but aren't used by the vast majority of people.

I can add this feature, but only after I see that a lot of people want the exact same implementation.

**Comment by creati0ns** [comment]
True, this specific referenced project is more a combination of multi site, with multi tenancy. However, the customers can not change views. it is more applicable to the multi site sites.

**Comment by rmsj** [comment]
Do you need help?
Keen to help. We are using previous version and already in production.
Only thing I've extended is the multiple credentials - one for each new database as I've talked to you on a separate issue.

**Comment by stancl** [comment]
Not really, the 3.x codebase is 99% done, I'm just doing some final changes and finishing the documentation. The only thing I'd appreciate would be some donations, I've been working on the v3 rewrite for about 4 weeks now and it's [biiig](https://github.com/stancl/tenancy/compare/2.x...master#files_bucket) 😅

You can also try upgrading to 3.x in a new branch of your application: https://www.notion.so/Upgrading-from-2-x-9a88ace0011b465e89e5fde05ea15887

**Comment by rmsj** [comment]
Cool. May I ask about the multiple database part of it? Did multiple credentials made the cut in this initial version?

**Comment by stancl** [comment]
Yeah, you can specify the DB user/password when creating the tenant, or you can let them be generated randomly (or you can use the same user for all DBs like in 2.x).

**Comment by rmsj** [comment]
Great!!! I will give it a go on a new project - and looking forward to upgrade current one to version 3.
Thanks for the hard work.

**Comment by LiXunGuan** [comment]
Great, half a month later, a new project is going to be developed with V3 version, how long will it work normally?

**Comment by dominikgeimer** [comment]
Do you have an ETA for the demo app?

**Comment by stancl** [comment]
1-3 weeks for sure, ideally within 2 weeks.

**Comment by enkota** [comment]
Very interested to read about **shared resource syncing** - Will this allow users on the central app  to be easily accessed from a tenant instance?

**Comment by stancl** [comment]
https://www.notion.so/Synced-resources-between-tenants-dbc3bcf221b44d0ab2ee7ed970542fe2

**Comment by enkota** [comment]
@stancl awesome thanks! Let's say we have a `Post` model on the tenant db but would like to reference the `user_id` foreign key constraint of that post with the central app `Users` db. Is this a better approach by syncing the users on the tenant db and reference `id` there?

**Comment by stancl** [comment]
Yeah, referencing stuff between databases is a bad idea. It can work (albeit not well with Eloquent), but it will break down the second you need a second database server.

**Comment by stancl** [comment]
v3.0.0-rc1 was tagged yesterday & the new website was published: https://tenancyforlaravel.com/

**Comment by vahidalvandi** [comment]
after create user in tenant in center db copy it with tenant id , it may duplicate to other tenant id

**Comment by stancl** [comment]
What?

**Comment by vahidalvandi** [comment]
> What?

i mistake , i create user in tenant sync  to center , with id in tenant , must first create in center and update in tenant

**Comment by iizno** [comment]
Is there any v3 equivalent to the v2 tenancy()->findByXXX() method ? I did not found anything in the documentation. 

If not, what is the best way to find a Tenant with a custom data named "Code" ?

A new code column in the tenants table ?

**Comment by stancl** [comment]
It’s a model.

```
Tenant::where(‘xxx’, ‘yyy’)->first();
// or
Tenant::firstWhere(‘xxx’, ‘yyy’);
```

Samuel Štancl

**Comment by iizno** [comment]
Ok, thanks !
But may I ask what is the purpose of the data column if I need to use new columns in the v3 for custom data ?

**Comment by stancl** [comment]
You don't need new columns for custom data. If you store something and a column doesn't exist for it, it will go to the data column.

Also, the package stores internal data, like DB names, credentials, etc. This way you don't have to create a column whenever the package starts using a new internal key.

**Comment by iizno** [comment]
Ok,so if I understand this correctly, I need to create a custom column only if I need to do queries on it. If it's only data for display / customisation, the data column is enough. That's it ?

**Comment by stancl** [comment]
You create custom columns when you need to improve performance.

You can query the data column too. MySQL can do that.

```php
Tenant::where('data.foo', 'bar')
```

**Comment by iizno** [comment]
Thanks, that's exactly what I needed, I didn't know that, thank you very much for your answers :)

**Comment by stancl** [comment]
🚀
